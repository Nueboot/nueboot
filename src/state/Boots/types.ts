import { BootInfo } from 'api/types';

export interface BootsState {
  readonly boots: BootInfo[];
  readonly loading: boolean;
}
