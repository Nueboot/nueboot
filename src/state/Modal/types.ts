export interface ModalState {
  modalType: ModalTypes | null;
}

export type ModalTypes = 'login' | 'signup' | 'reset';
