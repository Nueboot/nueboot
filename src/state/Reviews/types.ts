import { BootReview } from 'api/types';

export interface ReviewsState {
  readonly byId: { [key: string]: BootReview[] };
}
