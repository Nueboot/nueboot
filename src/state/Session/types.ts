export interface SessionState {
  error?: string;
  loggedIn: boolean;
  pending: boolean;
}

export interface LoginInfo {
  email: string;
  password: string;
}
