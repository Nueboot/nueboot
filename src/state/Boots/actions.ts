import { BootInfo } from 'api/types';
import { getAllBoots as fetchBoots, getBoot as fetchBoot } from '../../api';

export type FetchAllBoots = () => FetchingAllBootsAction;

export interface FetchingAllBootsAction {
  readonly type: 'BOOTS.FETCHING_BOOTS';
}

export const fetchingBoots: FetchAllBoots = () => ({
  type: 'BOOTS.FETCHING_BOOTS',
});

export type GetAllBootsSuccess = (
  boots: BootInfo[]
) => GetAllBootsSuccessAction;

export interface GetAllBootsSuccessAction {
  readonly type: 'BOOTS.RECEIVE_BOOTS';
  readonly payload: {
    boots: BootInfo[];
  };
}

export const getAllBootsSuccess: GetAllBootsSuccess = boots => ({
  type: 'BOOTS.RECEIVE_BOOTS',
  payload: { boots },
});

export const getAllBoots = () => dispatch => {
  dispatch(fetchingBoots());
  fetchBoots()
    .then(boots => {
      dispatch(getAllBootsSuccess(boots));
    })
    .catch(err => {
      return err;
    });
};

export interface GetBootAction {
  readonly type: 'BOOTS.RECEIVE_BOOT';
  readonly payload: {
    id: string;
    boot: BootInfo;
  };
}

export const getBootSuccess = (id: string, boot: BootInfo): GetBootAction => ({
  type: 'BOOTS.RECEIVE_BOOT',
  payload: {
    id,
    boot,
  },
});

export const getBoot = (id: string) => dispatch => {
  dispatch(fetchingBoots());
  fetchBoot(id)
    .then(boot => {
      dispatch(getBootSuccess(id, boot));
    })
    .catch(err => {
      throw err;
    });
};
