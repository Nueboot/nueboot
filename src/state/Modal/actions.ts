import { ModalTypes } from './types';

export interface OpenModalAction {
  readonly type: 'MODAL.OPEN_MODAL';
  readonly payload: { type: ModalTypes };
}

export type OpenModal = (type: ModalTypes) => OpenModalAction;

export const openModal: OpenModal = type => ({
  type: 'MODAL.OPEN_MODAL',
  payload: {
    type,
  },
});

export interface CloseModalAction {
  readonly type: 'MODAL.CLOSE_MODAL';
}

export type CloseModal = () => CloseModalAction;

export const closeModal: CloseModal = () => ({
  type: 'MODAL.CLOSE_MODAL',
});
