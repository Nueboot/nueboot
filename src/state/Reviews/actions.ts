import { addReview, fetchBootReview } from 'api';
import { BootReview } from 'api/types';

export const createReview = (review: BootReview, user: string) => dispatch => {
  addReview(review, user).then(dispatch(fetchReview(review.id)));
};

export const receiveReview = (reviews: BootReview[]) => ({
  type: 'REVIEWS.RECEIVE_REVIEW',
  payload: {
    ...reviews,
  },
});

export const fetchReview = (id: string) => dispatch => {
  fetchBootReview(id).then(reviews => {
    dispatch(receiveReview([reviews as BootReview]));
  });
};
