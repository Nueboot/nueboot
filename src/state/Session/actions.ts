import { auth, FirebaseError } from 'firebase';
import firebase from '../../lib/firebase';

export const verifyUser = () => dispatch => {
  dispatch(verifyingUser());
  firebase.auth().onAuthStateChanged(user => {
    if (user) {
      dispatch(loginSuccess());
    } else {
      dispatch(logoutUser());
    }
  });
};

export interface VerifyingUserAction {
  readonly type: 'SESSION.PENDING';
}

export const verifyingUser: () => VerifyingUserAction = () => ({
  type: 'SESSION.PENDING',
});

export interface LoginSuccessAction {
  readonly type: 'SESSION.LOGIN_SUCCESS';
}

export const loginSuccess: () => LoginSuccessAction = () => ({
  type: 'SESSION.LOGIN_SUCCESS',
});

export interface LoginErrorAction {
  readonly type: 'SESSION.ERROR';
  readonly payload: {
    error: FirebaseError,
  };
}

export const sessionError: (err: FirebaseError) => LoginErrorAction = error => ({
  payload: {
    error,
  },
  type: 'SESSION.ERROR',
});

export const logoutUser = () => dispatch => {
  try {
    firebase.auth().signOut();
    dispatch(logOutSuccess());
  } catch (err) {
    dispatch(sessionError(err.code));
  }
};

export interface LogOutSuccessAction {
  readonly type: 'SESSION.LOGOUT_SUCCESS';
}

export const logOutSuccess: () => LogOutSuccessAction = () => ({
  type: 'SESSION.LOGOUT_SUCCESS',
});

export const loginWithGoogle = () => dispatch => {
  const provider = new auth.GoogleAuthProvider();
  dispatch(oauthLogin(provider));
};

export const loginWithFacebook = () => dispatch => {
  const provider = new auth.FacebookAuthProvider();
  dispatch(oauthLogin(provider));
};

export const oauthLogin = provider => dispatch => {
  firebase.auth().signInWithPopup(provider).then(res => {
    dispatch(loginSuccess());
  }).catch(err => {
    dispatch(sessionError(err));
  });
};
