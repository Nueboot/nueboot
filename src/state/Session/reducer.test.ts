import { FirebaseError } from 'firebase';
import { CloseModalAction } from '../Modal/actions';
import { LoginErrorAction, LoginSuccessAction, LogOutSuccessAction, VerifyingUserAction } from './actions';
import sessionReducer from './reducer';
import { SessionState } from './types';

describe('Session reducer', () => {
  let state: SessionState;
  let newState: SessionState;
  let error: FirebaseError;

  beforeEach(() => {
    error = {
      name:  'Error',
      message:  'Error',
      code:  'Error',
    };
    state = {
      error: {
        name:  'Error',
        message:  'Error',
        code:  'Error',
      },
      loggedIn: false,
      pending: false,
    };
  });

  describe('with no action or state passed in', () => {
    beforeEach(() => {
      newState = sessionReducer();
    });

    it('returns the the default state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('successful login', () => {
    let action: LoginSuccessAction;
    beforeEach(() => {
      action = {
        type: 'SESSION.LOGIN_SUCCESS',
        payload: {
          userInfo: {},
        },
      };

      newState = sessionReducer(state, action);
    });

    it('returns the new state with loggedIn set to true', () => {
      expect(newState.loggedIn).toBe(true);
    });
  });

  describe('login error', () => {
    let action: LoginErrorAction;
    beforeEach(() => {
      action = {
        payload: {
          error,
        },
        type: 'SESSION.ERROR',
      };

      newState = sessionReducer(state, action);
    });

    it('returns the new state with the error object', () => {
      expect(newState.error).toEqual(error);
    });
  });

  describe('signout success', () => {
    let action: LogOutSuccessAction;
    beforeEach(() => {
      action = {
        type: 'SESSION.LOGOUT_SUCCESS',
      };

      newState = sessionReducer(state, action);
    });

    it('returns the new state with logged in set to false', () => {
      expect(newState.loggedIn).toBe(false);
    });
  });

  describe('verifying user', () => {
    let action: VerifyingUserAction;
    beforeEach(() => {
      action = {
        type: 'SESSION.PENDING',
      };

      newState = sessionReducer(state, action);
    });

    it('returns the new state with pending set to true', () => {
      expect(newState.pending).toBe(true);
    });
  });

  describe('close modal', () => {
    let action: CloseModalAction;
    beforeEach(() => {
      action = {
        type: 'MODAL.CLOSE_MODAL',
      };
      state = {
        ...state,
        error,
      };

      newState = sessionReducer(state, action);
    });

    it('clears the errors when closing the modal', () => {
      expect(newState.error).toBe(null);
    });
  });
});
