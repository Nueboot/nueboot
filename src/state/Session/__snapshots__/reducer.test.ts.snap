// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session reducer with no action or state passed in returns the the default state 1`] = `
Object {
  "error": null,
  "loggedIn": false,
  "pending": false,
  "user": Object {},
}
`;
